### ISSUE TITLE
Server initialization hangs if a plugin cannot be loaded

### ISSUE THREAD

**Issue by joshuaboniface** [jellyfin/jellyfin#4690]
**Describe the bug**
Server initialization hangs on startup if there is a plugin present that cannot be properly loaded. No errors are reported, the initialization just hangs indefinitely.

**To Reproduce**
1. Take a 10.6.4 release with the latest-at-the-time version of a plugin (I tested with `LDAP Authentication` version 9).
2. Upgrade the system to 10.7.0-rc1.
3. Restart the system. It will hang.

**Expected behavior**
Invalid plugins should just fail the load and not block startup. Perhaps display a Dashboard toast somewhere about this if it happens.

**Logs**
Logs courtesy of @ABotelho23 in #4676 where this was the underlying issue.

```
[1] Emby.Server.Implementations.ApplicationHost: Loading assemblies [2020-12-05 22:23:39.042 +00:00] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loaded assembly "LDAP-Auth, Version=9.0.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/LDAP Authentication/LDAP-Auth.dll" [2020-12-05 22:23:39.043 +00:00] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loaded assembly "Novell.Directory.Ldap.NETStandard, Version=3.2.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/LDAP Authentication/Novell.Directory.Ldap.NETStandard.dll" [2020-12-05 22:23:39.046 +00:00] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loaded assembly "OpenSubtitlesHandler, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/Open Subtitles/OpenSubtitlesHandler.dll" [2020-12-05 22:23:39.047 +00:00] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loaded assembly "Jellyfin.Plugin.OpenSubtitles, Version=9.0.0.0, Culture=neutral, PublicKeyToken=null" from "/config/plugins/Open Subtitles/Jellyfin.Plugin.OpenSubtitles.dll"
[hang]
```

My similar logs:

```
Dec 05 03:15:02 jf1.i.bonilan.net systemd[1]: Started Jellyfin Media Server.
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Jellyfin version: 10.7.0
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Environment Variables: ["[JELLYFIN_DATA_DIR, /srv/jellyfin]", "[JELLYFIN_ARGS, $JELLYFIN_WEB_OPT $JELLYFIN_RESTART_OPT $JELLYFIN_FFMPEG_OPT $JELLYFIN_SERVICE_OPT
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Arguments: ["/usr/lib/jellyfin/bin/jellyfin.dll", "--webdir=/usr/share/jellyfin/web", "--restartpath=/usr/lib/jellyfin/restart.sh", "--ffmpeg=/usr/share/jellyfin
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Operating system: Linux
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Architecture: X64
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: 64-Bit Process: True
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: User Interactive: True
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Processor count: 4
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Program data path: /srv/jellyfin
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Web resources path: /usr/share/jellyfin/web
Dec 05 03:15:03 jf1.i.bonilan.net jellyfin[13479]: [03:15:03] [INF] [1] Main: Application directory: /usr/lib/jellyfin/bin/
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Emby.Server.Implementations.AppBase.BaseConfigurationManager: Setting cache path: /srv/jellyfin/cache
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Defined LAN addresses : [127.0.0.1/32,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16]
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Defined LAN exclusions : []
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using LAN addresses: [127.0.0.1/32,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16]
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using bind addresses: []
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using bind exclusions: []
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loading assemblies
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loaded assembly LDAP-Auth, Version=9.0.0.0, Culture=neutral, PublicKeyToken=null from /srv/jellyfin/plugins/LDAP Authentic
Dec 05 03:15:04 jf1.i.bonilan.net jellyfin[13479]: [03:15:04] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loaded assembly Novell.Directory.Ldap.NETStandard, Version=3.2.0.0, Culture=neutral, PublicKeyToken=null from /srv/jellyfi
Dec 05 03:20:38 jf1.i.bonilan.net systemd[1]: Stopping Jellyfin Media Server...
Dec 05 03:20:38 jf1.i.bonilan.net jellyfin[13479]: [03:20:38] [INF] [2] Main: Received a SIGTERM signal, shutting down
Dec 05 03:20:38 jf1.i.bonilan.net systemd[1]: jellyfin.service: Main process exited, code=exited, status=143/n/a
Dec 05 03:20:38 jf1.i.bonilan.net systemd[1]: jellyfin.service: Failed with result 'exit-code'.
Dec 05 03:20:38 jf1.i.bonilan.net systemd[1]: Stopped Jellyfin Media Server.
```

**Screenshots**
N/A

**Additional context**
Probably related to DI changes, cc: @BaronGreenback

**Comment by BaronGreenback** [comment]
There are two "loading" instances, for the plugins. The initial is so that the plugin can register itself into DI. The second is when the plugin instance is created.
Both are contained with a try/catch block and will continue to run if the plugin fails, reporting the error in the log.

This look to be caused by something in the plugin itself. I'll go take a look.

**Comment by BaronGreenback** [comment]
Suspect the error is in RunStartupTasksAsync.

**Comment by BaronGreenback** [comment]
Error is Insufficient stack, is the cause of the blowout.

```
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(Microsoft.Extensions.DependencyInjection.ServiceLookup.ConstructorCallSite constructorCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context) Line 36
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs(36)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, System.__Canon>.VisitCallSiteMain(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext argument) Line 49
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteVisitor.cs(49)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngine, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverLock lockType) Line 92
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs(92)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite singletonCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context) Line 60
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs(60)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, object>.VisitCallSite(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext argument) Line 28
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteVisitor.cs(28)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope) Line 15
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs(15)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.RealizeService.AnonymousMethod__0(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope) Line 25
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/DynamicServiceProviderEngine.cs(25)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(System.Type serviceType, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngineScope) Line 92
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngine.cs(92)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(System.Type serviceType) Line 66
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngine.cs(66)
Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(System.Type serviceType) Line 58
	at /_/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs(58)
Microsoft.Extensions.DependencyInjection.Abstractions.dll!Microsoft.Extensions.DependencyInjection.ActivatorUtilities.ConstructorMatcher.CreateInstance(System.IServiceProvider provider) Line 408
	at /_/src/libraries/Common/src/Extensions/ActivatorUtilities/ActivatorUtilities.cs(408)
Microsoft.Extensions.DependencyInjection.Abstractions.dll!Microsoft.Extensions.DependencyInjection.ActivatorUtilities.CreateInstance(System.IServiceProvider provider, System.Type instanceType, object[] parameters) Line 93
	at /_/src/libraries/Common/src/Extensions/ActivatorUtilities/ActivatorUtilities.cs(93)
Emby.Server.Implementations.dll!Emby.Server.Implementations.ApplicationHost.CreateInstanceSafe(System.Type type) Line 399
	at C:\Code\jellyfin\Emby.Server.Implementations\ApplicationHost.cs(399)
System.Linq.dll!System.Linq.Enumerable.WhereSelectArrayIterator<System.Type, object>.MoveNext() Line 277
	at /_/src/libraries/System.Linq/src/System/Linq/Where.cs(277)
System.Linq.dll!System.Linq.Enumerable.WhereEnumerableIterator<object>.MoveNext() Line 117
	at /_/src/libraries/System.Linq/src/System/Linq/Where.cs(117)
System.Linq.dll!System.Linq.Enumerable.CastIterator<MediaBrowser.Controller.Authentication.IAuthenticationProvider>(System.Collections.IEnumerable source) Line 53
	at /_/src/libraries/System.Linq/src/System/Linq/Cast.cs(53)
System.Private.CoreLib.dll!System.Collections.Generic.List<MediaBrowser.Controller.Authentication.IAuthenticationProvider>.List(System.Collections.Generic.IEnumerable<MediaBrowser.Controller.Authentication.IAuthenticationProvider> collection) Line 85
	at /_/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs(85)
System.Linq.dll!System.Linq.Enumerable.ToList<MediaBrowser.Controller.Authentication.IAuthenticationProvider>(System.Collections.Generic.IEnumerable<MediaBrowser.Controller.Authentication.IAuthenticationProvider> source) Line 29
	at /_/src/libraries/System.Linq/src/System/Linq/ToCollection.cs(29)
Emby.Server.Implementations.dll!Emby.Server.Implementations.ApplicationHost.GetExports<MediaBrowser.Controller.Authentication.IAuthenticationProvider>(bool manageLifetime) Line 431
	at C:\Code\jellyfin\Emby.Server.Implementations\ApplicationHost.cs(431)
Jellyfin.Server.Implementations.dll!Jellyfin.Server.Implementations.Users.UserManager.UserManager(Jellyfin.Server.Implementations.JellyfinDbProvider dbProvider, MediaBrowser.Controller.Events.IEventManager eventManager, MediaBrowser.Model.Cryptography.ICryptoProvider cryptoProvider, MediaBrowser.Common.Net.INetworkManager networkManager, MediaBrowser.Common.IApplicationHost appHost, MediaBrowser.Controller.Drawing.IImageProcessor imageProcessor, Microsoft.Extensions.Logging.ILogger<Jellyfin.Server.Implementations.Users.UserManager> logger) Line 82
	at C:\Code\jellyfin\Jellyfin.Server.Implementations\Users\UserManager.cs(82)
```

**Comment by BaronGreenback** [comment]
Looks to be 

public async Task<ProviderAuthenticationResult> Authenticate(string username, string password)
{
         **var userManager = _applicationHost.Resolve<IUserManager>();**

If this is being called before initialization is complete, then it will cause the loop seen, as Usermanager depends upon IAuthenticationProviders which depends upon Usermanager.

The resolve would be better in the plugin.

**Comment by nielsvanvelzen** [comment]
Additional issue with plugin loading:

Locally installed plugins that are placed directly in the plugin folder root are not loaded anymore. In my case this caused my "resolver" plugin to not load and deleting my entire library.
Moving it to a subdirectory fixed  the problem, but this new behavior **must** be reverted or noted in our release notes.

This works:

```
jellyfin-data/
    plugins/
        test/
            resolver.dll
```

This does not (but did in 10.6):

```
jellyfin-data/
    plugins/
        resolver.dll
```

**Comment by BaronGreenback** [comment]
Didn't know plugins could stored in the root folder. It's going to have to go in the release notes.

This change was partially to do with manifests (unknown how you would be able to have multiple in one folder) and partially to get around the upgrading of plugins on the windows environment.

**Comment by BaronGreenback** [comment]
@crobibero  Can we get a rebuild of this plugin "re-published" to the repository. A local build based on the plugin code works. The repository doesn't.

**Comment by oddstr13** [comment]
@BaronGreenback Please note that manifests is something new from jprm, and their inclusion in the plugin zip was mostly for the convenience of managing packaged up plugins – I never intended for them to be required by the server.

**Comment by BaronGreenback** [comment]
system starts up plugin v10.

**Comment by joshuaboniface** [comment]
Yea the actual issue here is that incompatible 10.6.z plugins were hanging the startup; v10 works fine. So our workaround should probably just be to detect this and abort loading.

The root vs. folder thing is worth documenting but not a big deal I don't think, since it should only really affect manual plugin installs.

**Comment by joshuaboniface** [comment]
@oddstr13 Maybe not intended, but storing this metadata and having the server read it might be a good idea long-term for greater flexibility.

**Comment by cvium** [comment]
Fixed by #4709 and https://github.com/jellyfin/jellyfin/pull/5090
